function/datastore.js

```diff
@@ -38,42 +37,147 @@ const generateUserResponseKey = function(appId, tosVersion) {
     return generateKey(userResponseParts);
 }
 
-const getTOS = function(appid, tosversion) {
-    const query = datastore
+// define datastore clients as classes for ease of unit testing - unit tests can mock out parts of these classes.
+class GoogleDatastoreClient {
+
+    // ================================
+    // SUPPORT
+    // ================================
+
+    /**
+     * Generic query handler.
+     *
+     * Accepts error messages with the following keys:
+     *  0: if the query found no results
+     *  2: if the query found > 1 result
+     *
+     * Error message values are of the form {statusCode: 123, message: 'string or undefined'}
+     *
+     * @param {*} queryPromise Promise that contains datastore results
+     * @param {*} errorMessages map of error messages to handle
+     */
+    resultHandler(queryPromise, errorMessages) {
+        return queryPromise
+            .then( results => {
+                // results object is an array that contains [0]: array of rows returned; [1]: metadata about the results
+                const hits = results[0];
+                if (hits.length === 1) {
+                    return hits[0];
+                } else if (hits.length == 0) {
+                    return rejection(errorMessages[0].statusCode, errorMessages[0].message);
+                } else {
+                    // defensive! This should never happen since all known code either sets limit(1) in its query
+                    // or is querying for a specific record id.
+                    return rejection(errorMessages[2].statusCode, errorMessages[2].message);
+                }
+            })
+            .catch( err => {
+                if (err.statusCode) {
+                    return rejection(err.statusCode, err);
+                } else {
+                    return rejection(500, err);
```

Comment:
I originally had it without catch clauses here for exactly this goal, and ran into problems - my problems must have been elsewhere, because you're absolutely right they're not needed!